src/utils/hooks/useAsyncActionHandler.ts

```diff
@@ -1,58 +1,67 @@
+ */
 export const useAsyncActionHandler = () => {
   const [isLoading, setIsLoading] = useState(false);
+  const isLoadingRef = useRef(isLoading);
   const toast = useToast();
   const dispatch = useAppDispatch();
 
-  // wraps an async function that might throw an error (e.g. network call, decryption, raw input parsing, etc.)
-  // handles `isLoading` state and sets it back to `false` after the function is done
-  //
-  // if an exception was thrown it'll show a toast with the error message and re-throws and dispatches the error to redux
-  // (might be needed to additionally handle the error in the caller function)
-  //
-  // helps to prevent multiple clicks on a button that triggers an async function
-  // (though it's most likely is still vulnerable to race conditions)
-  const handleAsyncActionUnsafe = async <T>(
-    fn: () => Promise<T>,
-    toastOptions?: UseToastOptions | ((error: any) => UseToastOptions)
-  ): Promise<T | void> => {
-    if (isLoading) {
-      return;
-    }
-    setIsLoading(true);
-    try {
-      return await fn();
-    } catch (error: any) {
-      const errorContext = getErrorContext(error);
+  const handleAsyncActionUnsafe = useCallback(
+    async <T>(
+      fn: () => Promise<T>,
+      toastOptions?: UseToastOptions | ((error: any) => UseToastOptions)
+    ): Promise<T | void> => {
+      if (isLoadingRef.current) {
+        return;
+      }
+      isLoadingRef.current = true;
+      setIsLoading(true);
+      try {
+        return await fn();
+      } catch (error: any) {
+        const errorContext = getErrorContext(error);
 
-      toast({
-        description: errorContext.description,
-        status: "error",
-        ...(typeof toastOptions === "function" ? toastOptions(error) : toastOptions),
-      });
+        toast({
+          description: errorContext.description,
+          status: "error",
+          ...(typeof toastOptions === "function" ? toastOptions(error) : toastOptions),
+        });
 
-      dispatch(errorsSlice.actions.add(errorContext));
-      throw error;
-    } finally {
-      setIsLoading(false);
-    }
-  };
+        dispatch(errorsSlice.actions.add(errorContext));
+        throw error;
+      } finally {
+        isLoadingRef.current = false;
+        setIsLoading(false);
+      }
+    },
+    [dispatch, toast]
+  );
 
-  /**
-   * Same as {@link handleAsyncActionUnsafe}, but returns undefined instead of throwing an error.
-   *
-   * NOTE: might be helpful in tests to remove the catch part of this function so that you'd see all the exceptions.
-   *  */
-  const handleAsyncAction = async <T>(
-    fn: () => Promise<T>,
-    toastOptions?: UseToastOptions | ((error: any) => UseToastOptions)
-  ): Promise<T | void> => handleAsyncActionUnsafe(fn, toastOptions).catch(() => {});
+  // NOTE: might be helpful in tests to remove the catch part of this function so that you'd see all the exceptions.
+  const handleAsyncAction = useCallback(
```

Comment:
There is useCallback for `handleAsyncActionUnsafe`, do we need another one here?

Reply:
yes, because otherwise we'd keep creating new functions on each render & we wouldn't be able to use it as one of useEffect's dependencies